src/foobar/api.py

```diff
@@ -116,10 +116,67 @@ def list_purchases(account_id, start=None, stop=None, **kwargs):
 
     Together with each purchase, a list of purchased items is returned.
     """
-
     account_obj = Account.objects.get(id=account_id)
     purchase_objs = account_obj.purchases.filter(
         account_id=account_id,
         **kwargs
     )[start:stop]
     return [get_purchase(obj.id) for obj in purchase_objs]
+
+
+@transaction.atomic
+def calculate_correction(new_balance, owner_id, user, reference=None):
```

Comment:
Hmm the `reference` should be saved only in `WalletCorrection`. Then `WalletCorrect.id` should be saved as reference in `WalletTransaction`. That way, WalletTransactions can be traced back to specific WalletCorrections.